mlflow/tracing/export/mlflow.py

```diff
@@ -0,0 +1,146 @@
+
+_logger = logging.getLogger(__name__)
+
+
+class MLflowSpanExporter(SpanExporter):
+    """
+    An exporter implementation that logs the traces to MLflow.
+
+    MLflow backend (will) only support logging the complete trace, not incremental updates
+    for spans, so this exporter is designed to aggregate the spans into traces in memory.
+    Therefore, this only works within a single process application and not intended to work
+    in a distributed environment. For the same reason, this exporter should only be used with
+    SimpleSpanProcessor.
+
+    If we want to support distributed tracing, we should first implement an incremental trace
+    logging in MLflow backend, then we can get rid of the in-memory trace aggregation.
+    """
+
+    _MAX_CHARS_IN_TRACE_INFO_INPUTS_OUTPUTS = 300  # TBD
+
+    def __init__(self, client: TraceClient):
+        self._client = client
+        self._trace_aggregator = InMemoryTraceDataAggregator.get_instance()
+
+    def export(self, spans: Sequence[MLflowSpanWrapper]):
+        """
+        Export the spans to MLflow backend.
+
+        Args:
+            spans: A sequence of MLflowSpanWrapper objects to be exported. The base
+                OpenTelemetry (OTel) exporter should take OTel spans but this exporter
+                takes the wrapper object, so we can carry additional MLflow-specific
+                information such as inputs and outputs.
+        """
+        mlflow_spans = []
+        for span in spans:
+            if not isinstance(span, MLflowSpanWrapper):
+                _logger.warning(
+                    "Span exporter expected MLflowSpanWrapper, but got "
+                    f"{type(span)}. Skipping the span."
+                )
+            mlflow_span = span._to_mlflow_span()
+            self._trace_aggregator.add_span(mlflow_span)
+            mlflow_spans.append(mlflow_span)
+
+        # Call this after processing all spans because the parent-child order might
+        # not be preserved in the input spans
+        for span in mlflow_spans:
+            if span.parent_span_id is None:
+                self._export_trace(span)
+
+    def _export_trace(self, root_span: Span):
+        trace_data = self._trace_aggregator.pop_trace(root_span.context.trace_id)
+        if trace_data is None:
+            _logger.warning(f"Trace data with ID {root_span.context.trace_id} not found.")
+            return
+
+        # Create a TraceInfo object from the root span information
+        trace_info = TraceInfo(
+            trace_id=root_span.context.trace_id,
+            name=root_span.name,
+            start_time=root_span.start_time,
+            end_time=root_span.end_time,
+            status=root_span.status,
+            inputs=self._serialize_inputs_outputs(root_span.inputs),
+            outputs=self._serialize_inputs_outputs(root_span.outputs),
+            # TODO: These fields should be set if necessary
+            metadata={},
+            tags={},
```

Comment:
Q: Where are we going to get these if we need to set them? They're not part of spans, will we add more params to this 'export' function? Or from client?

Reply:
Hmm great point, yea it can't be fetched from span.
There can be a few ways to set metadata.

* Option 1: Add a new API like `mlflow.set_trace_source()` to globally set model_id, app_version_id, etc. Similarly to how we can fetch run_id from global active_run object.
* Option 2: Mandate users to decorate the predict function with @mlflow.trace when they want those metadata. This way we can inspect them from `self` argument and pass them via MLflowSpanWrapper object.
* Option 3: Add an backend API for updating the trace after logged (the one discussed in the design doc earlier).

None of these is a perfect option, so need design decision... Let me update the design doc and ask more opinions.

Reply:
Added a section in [the doc](https://docs.google.com/document/d/15iOQicMz8hP6-VsVLs_l26TrDMNoVd_zb-tWRcTUtgo/edit#heading=h.k5g8vy7zzbhf). Once we settle on one of the approaches, I will implement it in a follow-up.